protocols/request-response/src/handler/protocol.rs

```diff
@@ -0,0 +1,161 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! The definition of a request/response protocol via inbound
+//! and outbound substream upgrades. The inbound upgrade
+//! receives a request and sends a response, whereas the
+//! outbound upgrade send a request and receives a response.
+
+use crate::RequestId;
+use crate::codec::RequestResponseCodec;
+
+use futures::{
+    channel::oneshot,
+    future::BoxFuture,
+    prelude::*,
+};
+use libp2p_core::{
+    upgrade::{InboundUpgrade, OutboundUpgrade, UpgradeInfo},
+};
+use libp2p_swarm::{
+    NegotiatedSubstream,
+};
+use smallvec::SmallVec;
+use std::io;
+
+/// The level of support for a particular protocol.
+#[derive(Debug, Clone)]
+pub enum ProtocolSupport {
+    /// The protocol is only supported for inbound requests.
+    Inbound,
+    /// The protocol is only supported for outbound requests.
+    Outbound,
+    /// The protocol is supported for inbound and outbound requests.
+    Full
+}
+
+impl ProtocolSupport {
+    /// Whether inbound requests are supported.
+    pub fn inbound(&self) -> bool {
+        match self {
+            ProtocolSupport::Inbound | ProtocolSupport::Full => true,
+            ProtocolSupport::Outbound => false,
+        }
+    }
+
+    /// Whether outbound requests are supported.
+    pub fn outbound(&self) -> bool {
+        match self {
+            ProtocolSupport::Outbound | ProtocolSupport::Full => true,
+            ProtocolSupport::Inbound => false,
+        }
+    }
+}
+
+/// Response substream upgrade protocol.
+///
+/// Receives a request and sends a response.
+#[derive(Debug)]
+pub struct ResponseProtocol<TCodec>
+where
+    TCodec: RequestResponseCodec
+{
+    pub(crate) codec: TCodec,
+    pub(crate) protocols: SmallVec<[TCodec::Protocol; 2]>,
+    pub(crate) request_sender: oneshot::Sender<TCodec::Request>,
+    pub(crate) response_receiver: oneshot::Receiver<TCodec::Response>
+}
+
+impl<TCodec> UpgradeInfo for ResponseProtocol<TCodec>
+where
+    TCodec: RequestResponseCodec
+{
+    type Info = TCodec::Protocol;
+    type InfoIter = smallvec::IntoIter<[Self::Info; 2]>;
+
+    fn protocol_info(&self) -> Self::InfoIter {
+        self.protocols.clone().into_iter()
+    }
+}
+
+impl<TCodec> InboundUpgrade<NegotiatedSubstream> for ResponseProtocol<TCodec>
+where
+    TCodec: RequestResponseCodec + Send + Sync + 'static,
+{
+    type Output = ();
```

Comment:
This is motivated [here](https://github.com/libp2p/rust-libp2p/blob/0b96c7851782b97be5bae1b6fd7801f1b1883d3d/protocols/request-response/src/handler.rs#L157-L162). In essence, it makes for a much simpler implementation (I had the obvious one you suggest before) for a variety of reasons:

  1) The protocols handler does not need to track pending inbound substreams as well as the subsequent sending of responses, e.g. via a dedicated `FuturesUnordered`.
  2) The protocols handler does not require another command for sending responses back from the behaviour to the handler and thus `TResponse: Clone` as well.
  3) The protocols handler does not need to track timeouts for pending responses to inbound substreams - the `InboundUpgrade` already has a timeout managed by the Swarm, just like the `OutboundUpgrade`. Otherwise, without timeouts after the request has "left" the `InboundUpgrade`, and the handler itself tracking inbound substreams around while it is waiting to get a response, a single omission of `RequestResponse::send_response` would result in a memory leak in the handler. Not only that but being forced to call `send_response` even for one-way protocols is inconvenient from the perspective of the API. All of these problems are avoided by keeping the read-request/send-response I/O inside the `InboundUpgrade::apply_upgrade`.
  4) Lastly, but certainly not most importantly, there is the symmetry: `OutboundUpgrade` (send request -> read response); `InboundUpgrade` (read request -> send response).

Reply:
I do not question the relative simplicity, but would you not agree that this deviates from how `ProtocolHandler`s are intended to work? If it turns out that they introduce unnecessary complexity, does this not indicate that their role needs to be revisited?

Reply:
> I do not question the relative simplicity, but would you not agree that this deviates from how ProtocolHandlers are intended to work?

I honestly don't know about whether there is a specific way in which they are intended to work and thus how much this deviates. I chose this implementation purely based on its perceived benefits for the implementation of request/response-per-substream protocols in particular. In the general case, I don't think it is "normal" to send application-level requests/responses in `Inbound/Outbound` upgrades at all, just to negotiate protocols. In that sense I don't even think it is intended behaviour for an `InboundUpgrade` to even yield a request.

>  If it turns out that they introduce unnecessary complexity, does this not indicate that their role needs to be revisited?

I agree in general. I think there was actually some loose agreement at some point on trying to remove the notion of a `ProtocolsHandler` altogether, possibly alongside rethinking the `Swarm`, but I don't think anyone has seriously made an attempt to do that so far.

Reply:
> At the time when these traits were created, they were indeed only supposed to negotiate protocols. However the idea of using them for application-level requests/responses is the natural continuation of this. This pattern is already used everywhere, and I think it is indeed a better pattern compared to just negotiating protocols and handling the application-level logic in the ProtocolsHandler. The only problem to me is that the names InboundUpgrade and OutboundUpgrade do not reflect this.

That is what I thought, thanks for the clarification.